### ISSUE TITLE
[MNT]: Ease building just one Sphinx Gallery example file

### ISSUE THREAD

**Issue by AlexSzatmary** [matplotlib/matplotlib#28042]
### Summary

Someone writing a Sphinx Gallery example needs to wait a long time for it to build through `make html` or do `make html-skip-subdirs` (which requires editing `.mpl_skip_subdirs.yaml` and still leads to a lot of unhhelpful warnings).

### Proposed fix

It isn't easy to figure out how to efficiently rebuild a single example. This use case is important because people developing gallery entries are focused on a single file and have to preview it many times. Also, writing a gallery entry can be a good way for beginners to enter a project but beginners are likely to struggle—but making this experience better will help projects gain contributors. A tweak to the docs would have made things a lot easier for me, and some adjustment to how Sphinx Gallery gets configured might make a better process automatic.

The Sphinx Gallery [docs](https://sphinx-gallery.github.io/stable/configuration.html) helpfully point out that you can build a single do 
```
sphinx-build -D sphinx_gallery_conf.filename_pattern=plot_specific_example\.py ...
```
but I found setting include_patterns to also be essential for that goal.

I added the following to conf.py:
```
include_patterns = ["index.rst",
"users/explain/axes/legend_guide.rst",
]
suppress_warnings = ["toc.not_readable", "ref.doc", "ref.obj", "ref.func", "ref.meth",
"ref.class"]
gallery_dirs = ["users/explain/axes"]
```
and run with `python -msphinx -b html -D sphinx_gallery_conf.filename_pattern=../galleries/users_explain/axes/legend_guide.py . build/html`

Each of those tweaks to conf.py was important.
* include_patterns was needed to stop Sphinx from spending a lot of time on unrelated files. Without having the rst file being included, the html never gets built.
* suppress_warnings is nice to add because, while it's understandable that building only one file would trigger a lot of warnings, it's helpful to turn them off so that the relevant ones stand out.
* gallery_dirs drops the build time for a single file by a few seconds, which is nice.

1. I'd like the docs to give these hints on how to build a single file.
2. We could add another build target in the doc Makefile, which would trigger the relevant logic within conf.py.

I'm willing and able to take on both of these tasks and would like to confirm that this approach makes sense before proceeding.

**Comment by timhoffm** [comment]
Thanks for the idea! Making doc builds faster is highly desirable.

My first impression:
- I'm sceptical on 1.: Users modifying `conf.py` heavily is cumbersome and a recipe for messing up.
- 2. sounds viable

but: Is this the right solution? Even a fast doc build is a disruption. I would recommend to edit the example in VS Code cell mode (or similar tooling), because that gives you immediate feedback. When done one or two doc build runs should suffice to see how it finally looks in HTML.

On the technical side: I'm wondering though, why we have to specify the gallery example file. Granted, the first build is slow, but since sphinx-gallery only reruns stale builds, shouldn't all subsequent builds be approximately equally fast with and without specifying that file? What am I missing? Is the stale check really that time-consuming? Not having to specify the file is nice in terms of usability. And if it was not that much faster compared to auto-detection, not having to learn that variant and copy the filename to the sphinx call is also a value.

**Comment by AlexSzatmary** [comment]
Good catch, @timhoffm , on your final paragraph—I just checked and it's totally unnecessary. I had thought it was important because it was what the Sphinx Gallery docs pointed to. I have raised a similar issue with Sphinx Gallery. I think that the more critical fix would be on their end but there could be some configuration tweaks that could rapidly be implemented by Matplotlib.

include_patterns is the critical thing to set, it can be set easily at the command line, but unfortunately it's a headache for someone new to Sphinx and Sphinx Gallery to figure out that you need to specify a particular rst file as the include (because you don't directly edit that file).

And thanks for the tip on using cell mode. I'll try that, too.

**Comment by AlexSzatmary** [comment]
I'd like to put together a tool for this issue. Although cell mode works, building a single Sphinx gallery file should also work, and others would like it to work. Based on conversations on Elements, several others have tried to get this working but have not succeeded. The best current solution is html-skip-subdirs which requires a yaml file that excludes everything you're not working on, which leads to a workable solution but is cumbersome.

I plan to put something together—if only for my own use—and I'd like input from others as I do so.
